pkg/bosh/manifest/container_factory.go

```diff
@@ -51,22 +51,7 @@ func (c *ContainerFactory) JobsToInitContainers(jobs []Job, hasPersistentDisk bo
 		}
 
 		// Setup the BOSH pre-start init containers.
-		boshPreStart := filepath.Join(VolumeJobsDirMountPath, job.Name, "bin", "pre-start")
-		boshPreStartInitContainer := corev1.Container{
-			Name:         fmt.Sprintf("bosh-pre-start-%s", job.Name),
-			Image:        jobImage,
-			VolumeMounts: instanceGroupVolumeMounts(),
-			Command:      []string{"/bin/sh", "-c"},
-			Args:         []string{fmt.Sprintf(`if [ -x "%[1]s" ]; then "%[1]s"; fi`, boshPreStart)},
-		}
-		if hasPersistentDisk {
-			persistentDisk := corev1.VolumeMount{
-				Name:      VolumeStoreDirName,
-				MountPath: VolumeStoreDirMountPath,
-			}
-			boshPreStartInitContainer.VolumeMounts = append(boshPreStartInitContainer.VolumeMounts, persistentDisk)
-		}
-		boshPreStartInitContainers = append(boshPreStartInitContainers, boshPreStartInitContainer)
+		boshPreStartInitContainers = append(boshPreStartInitContainers, boshPreStartInitContainer(job.Name, jobImage, hasPersistentDisk))
```

Comment:
I'm not comfortable with all these memory copying. We are passing structs all around by value. On one hand, it prevents garbage collection from happening, on the other hand, it can get wild with structs growing. IMO we either leave the way it was or we start passing the pointers around.

Reply:
Yes the function call returns a struct, but I don't see a need to optimize this. The number of elements is rather small and it's not that costly.
 
More importantly, I think it's totally worth passing the struct for readability, since pointers signal a mutable relationship to the reader.  We also avoid dealing with nil.

https://medium.com/@vCabbage/go-are-pointers-a-performance-optimization-a95840d3ef85

If you don't put the struct into the HEAP, there is no need for garbage collection. I wouldn't use pointers for performance reasons (well, unless it's performance critical and I benchmarked first).

If you think it's really hurting us, please provide a benchmark and add a separate story to optimize the container factory.

Reply:
I don't think it's really hurting us at this point. It could though. I'll leave it for the future.